Simkin modular merge

alfredsimkin opened the pull request:
made small modifications to schema.json to facilitate MIP files
wrapped sections of merge into functions
added an 'init' block
changed output folder from 'PMO' to 'finished_PMO_files'

I didn't wrap the functions into classes - I can do this if it would be helpful for consistency with the other pages (left it this way because I think readability is better, but maybe there are advantages to implementing as classes that I'm unfamiliar with - I'd like to learn more about it)

kathryn1995 commented:
This looks so much neater, thanks Alfred! 

When I build the app I am seeing these errors when the panel info and bioinformatics info aren't entered yet:
<img width="1428" alt="Screenshot 2025-02-20 at 3 37 12 PM" src="https://github.com/user-attachments/assets/85442612-3d67-4a0c-9ae4-d1a6d3d8e05a" />
<img width="1435" alt="Screenshot 2025-02-20 at 3 38 06 PM" src="https://github.com/user-attachments/assets/123a7a6f-7056-41d5-b695-1b3b801f3994" />

I'm also wondering if the check functions could be made into one more general function as they have the same pattern?

kathryn1995 commented:
Also, what formatter are you using for your code? We seem to have some conflicts between what we're using

alfredsimkin commented on pages/8_Create_Final_PMO.py:
```diff
@@ -3,94 +3,125 @@
+        st.write(
+            "**Current Microhaplotype Information from bioinformatics run:**", bioinfo_id)
+    else:
+        bioinfo_id=None
```
fixes missing return value errors

alfredsimkin commented on pages/8_Create_Final_PMO.py:
```diff
@@ -3,94 +3,125 @@
+        panel_id = ', '.join(panel_info["panel_info"].keys())
+        st.write("**Current Panel Information:**", panel_id)
+    else:
+        panel_info, panel_id=None, None
```
fixes empty return value errors when streamlit sees that this data is missing

alfredsimkin commented on pages/8_Create_Final_PMO.py:
```diff
@@ -3,94 +3,125 @@
-        "No sequencing information found. Please go to the Sequencing Information tab before proceeding.")
+def merge_data(panel_info, panel_id, bioinfo_id):
+    # MERGE DATA
+    if ('panel_info' in st.session_state
```
Now the 'merge data' subheader and button only appear if all the other checks work - should probably move this to the 'main' block since all these checks are performed earlier

alfredsimkin commented:
> This looks so much neater, thanks Alfred!
> 
> When I build the app I am seeing these errors when the panel info and bioinformatics info aren't entered yet: <img alt="Screenshot 2025-02-20 at 3 37 12 PM" width="1428" src="https://private-user-images.githubusercontent.com/52662905/415415744-85442612-3d67-4a0c-9ae4-d1a6d3d8e05a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA4MDQ1OTIsIm5iZiI6MTc0MDgwNDI5MiwicGF0aCI6Ii81MjY2MjkwNS80MTU0MTU3NDQtODU0NDI2MTItM2Q2Ny00YTBjLTlhZTQtZDFhNmQzZDhlMDVhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzAxVDA0NDQ1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYzM2Q3YzY5ODk5OTEyZGQ3YjRhZTEyMGE5ZWRiZTlmNjNiMmQ3YmM3MmVmYzhkYWJiY2ZkN2I4NWVhNzBkY2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.B7zM1Ps3_JR48BdMLQKcTlnygomrgCpCiUKizFkwpZ0"> <img alt="Screenshot 2025-02-20 at 3 38 06 PM" width="1435" src="https://private-user-images.githubusercontent.com/52662905/415415908-123a7a6f-7056-41d5-b695-1b3b801f3994.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA4MDQ1OTIsIm5iZiI6MTc0MDgwNDI5MiwicGF0aCI6Ii81MjY2MjkwNS80MTU0MTU5MDgtMTIzYTdhNmYtNzA1Ni00MWQ1LWI2OTUtMWIzYjgwMWYzOTk0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzAxVDA0NDQ1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFiYWJjYTNmZDk4NThiMmFlZGZkOWVlZDM3NDI2MmVlZTdkZWNkZjA3MGVlYzdmYjVhOWFlNDRjM2I0ODM2ZWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.L0zxMYE7uA79j6bvC50cEy165A81EZ0NvUgaC8Pgx7Q">
> 
> I'm also wondering if the check functions could be made into one more general function as they have the same pattern?

Fixed this issue in latest commit - I like the idea of merging check functions into one - I would just need to standardize the messages upon 'success' somewhat. I'll experiment with this soon.

alfredsimkin commented:
> Also, what formatter are you using for your code? We seem to have some conflicts between what we're using

I'm using sublime - I usually use tabs instead of spaces for indentation. I converted the tabs back to spaces (I think) in latest commit

kathryn1995 commented:
Hi Alfred, I prefer the output from this one. I think it's more useful for the user to get all the things they've missed in one go. I was wondering more if the duplicated code across functions could be generalised into one that is then called for each key that needs to be checked. e.g. could check_specimen_info be converted to something like this and used for multiple fields instead of having a separate function for each 

```

def check_field_info(field_name, field, max_display):
    if field in st.session_state:
        field_values = ', '.join(
            list(st.session_state[field].keys())[:max_display])
        st.write(f"**Current Specimen Information for samples (showing first {max_display}):** ",
                 field_values, '...')
    else:
        st.error(
            f"No {field} found. Please go to the f{field_name} tab before proceeding.")
        
```

This is still much neater and the errors I had before are gone. So happy if you want to merge this or do something like the above.

alfredsimkin commented:
> Hi Alfred, I prefer the output from this one. I think it's more useful for the user to get all the things they've missed in one go. I was wondering more if the duplicated code across functions could be generalised into one that is then called for each key that needs to be checked. e.g. could check_specimen_info be converted to something like this and used for multiple fields instead of having a separate function for each
> 
> ```
> 
> def check_field_info(field_name, field, max_display):
>     if field in st.session_state:
>         field_values = ', '.join(
>             list(st.session_state[field].keys())[:max_display])
>         st.write(f"**Current Specimen Information for samples (showing first {max_display}):** ",
>                  field_values, '...')
>     else:
>         st.error(
>             f"No {field} found. Please go to the f{field_name} tab before proceeding.")
>         
> ```
> 
> This is still much neater and the errors I had before are gone. So happy if you want to merge this or do something like the above.

This function that you have above looks good, except it only works for functions for which 'field' is the top layer of streamlit.sessions() (i.e. specimen info and experiment info). By contrast, for panel_info, experiment_info, and microhaplotype_info there's an additional layer of nesting with hardcoded keys underneath the streamlit.sessions keys, and for the remaining functions there's an additional layer of nesting with user-defined keys under the streamlit.sessions keys. The 'specimen Information' content of the function would also need to be assigned to a variable. I went back to the other branch (modular_merge_unified_check_function) which uses a function that is very similar to the function you listed above. I removed the break statement in my function - this allows a user to see all the pages that 'passed' vs. all the ones that 'failed' in one go. I also modified streamlit.write to be streamlit.success and modified a duplicated subheader to accurately describe the demultiplexing page.
